### ISSUE TITLE
[Bug]: [mmap][benchmark] Turn on mmap, insert 16m 128dimensional data into a standalone 8c16g server, milvus reboots and load fails.

### ISSUE THREAD

**Issue by elstic** [milvus-io/milvus#25562]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version:master-20230713-b22d6d2a 
- Deployment mode(standalone or cluster):standalone 
- MQ type(rocksmq, pulsar or kafka):    rocksmq
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

argo task : fouramf-xjg5d

server  (milvus restart):
```
fouramf-xjg5d-77-4508-etcd-0                                      1/1     Running     0               72m     10.104.15.225   4am-node20   <none>           <none>
fouramf-xjg5d-77-4508-milvus-standalone-6d4d7dc698-4zq46          1/1     Running     1 (2m42s ago)   72m     10.104.9.107    4am-node14   <none>           <none>
fouramf-xjg5d-77-4508-minio-8cff4565b-dq89h                       1/1     Running     0               72m     10.104.15.224   4am-node20   <none>           <none>
```

client log:
```
[2023-07-13 06:51:05,962 -  INFO - fouram]: [Base] Start flush collection fouram_17Z9JGsc (base.py:278)
[2023-07-13 06:51:17,775 -  INFO - fouram]: [Base] Params of index: [{'float_vector': {'index_type': 'HNSW', 'metric_type': 'L2', 'params': {'M': 8, 'efConstruction': 200}}}] (base.py:442)
[2023-07-13 06:51:17,776 -  INFO - fouram]: [Base] Start release collection fouram_17Z9JGsc (base.py:289)
[2023-07-13 06:51:17,857 -  INFO - fouram]: [Base] Start build index of HNSW for collection fouram_17Z9JGsc, params:{'index_type': 'HNSW', 'metric_type': 'L2', 'params': {'M': 8, 'efConstruction': 200}} (base.py:428)
[2023-07-13 07:29:23,398 -  INFO - fouram]: [Time] Index run in 2285.5388s (api_request.py:45)
[2023-07-13 07:29:23,398 -  INFO - fouram]: [CommonCases] RT of build index HNSW: 2285.5388s (common_cases.py:96)
[2023-07-13 07:29:23,399 -  INFO - fouram]: [Base] Params of index: [{'float_vector': {'index_type': 'HNSW', 'metric_type': 'L2', 'params': {'M': 8, 'efConstruction': 200}}}] (base.py:442)
[2023-07-13 07:29:23,400 -  INFO - fouram]: [CommonCases] Prepare index HNSW done. (common_cases.py:99)
[2023-07-13 07:29:23,400 -  INFO - fouram]: [CommonCases] No scalars need to be indexed. (common_cases.py:107)
[2023-07-13 07:29:23,401 -  INFO - fouram]: [Base] Number of vectors in the collection(fouram_17Z9JGsc): 16000000 (base.py:469)
[2023-07-13 07:29:23,402 -  INFO - fouram]: [Base] Start load collection fouram_17Z9JGsc,replica_number:1,kwargs:{} (base.py:284)
[2023-07-13 07:29:44,716 - WARNING - fouram]: [93m[get_loading_progress] retry:4, cost: 0.27s, reason: <_MultiThreadedRendezvous: StatusCode.UNAVAILABLE, failed to connect to all addresses; last error: UNKNOWN: ipv4:10.255.147.216:19530: Failed to connect to remote host: Connection refused>[0m (decorators.py:71)
[2023-07-13 07:29:44,988 - WARNING - fouram]: [93m[get_loading_progress] retry:5, cost: 0.81s, reason: <_MultiThreadedRendezvous: StatusCode.UNAVAILABLE, failed to connect to all addresses; last error: UNKNOWN: ipv4:10.255.147.216:19530: Failed to connect to remote host: Connection refused>[0m (decorators.py:71)
[2023-07-13 07:29:45,799 - WARNING - fouram]: [93m[get_loading_progress] retry:6, cost: 2.43s, reason: <_MultiThreadedRendezvous: StatusCode.UNAVAILABLE, failed to connect to all addresses; last error: UNKNOWN: ipv4:10.255.147.216:19530: Failed to connect to remote host: Connection refused>[0m (decorators.py:71)
[2023-07-13 07:29:48,235 - WARNING - fouram]: [93m[get_loading_progress] retry:7, cost: 7.29s, reason: <_MultiThreadedRendezvous: StatusCode.UNAVAILABLE, failed to connect to all addresses; last error: UNKNOWN: ipv4:10.255.147.216:19530: Failed to connect to remote host: Connection refused>[0m (decorators.py:71)
[2023-07-13 07:29:55,547 - WARNING - fouram]: [93m[get_loading_progress] retry:8, cost: 21.87s, reason: <_MultiThreadedRendezvous: StatusCode.UNAVAILABLE, failed to connect to all addresses; last error: UNKNOWN: ipv4:10.255.147.216:19530: Failed to connect to remote host: Connection refused>[0m (decorators.py:71)
[2023-07-13 07:30:17,447 - WARNING - fouram]: [93m[get_loading_progress] retry:9, cost: 60.00s, reason: <_MultiThreadedRendezvous: StatusCode.UNAVAILABLE, failed to connect to all addresses; last error: UNKNOWN: ipv4:10.255.147.216:19530: Failed to connect to remote host: Connection refused>[0m (decorators.py:71)
[2023-07-13 07:31:17,509 - WARNING - fouram]: [93m[get_loading_progress] retry:10, cost: 60.00s, reason: <_MultiThreadedRendezvous: StatusCode.UNAVAILABLE, failed to connect to all addresses; last error: UNKNOWN: ipv4:10.255.147.216:19530: Failed to connect to remote host: Connection refused>[0m (decorators.py:71)
[2023-07-13 07:32:17,571 - WARNING - fouram]: [93mRetry run out of 10 retry times[0m (decorators.py:79)
[2023-07-13 07:32:17,571 - ERROR - fouram]: RPC error: [get_loading_progress], <MilvusException: (code=1, message=Retry run out of 10 retry times, message=collection 442821320644231354 has not been loaded to memory or load failed)>, <Time:{'RPC start': '2023-07-13 07:29:44.581672', 'RPC error': '2023-07-13 07:32:17.571654'}> (decorators.py:108)
[2023-07-13 07:32:17,572 - ERROR - fouram]: RPC error: [wait_for_loading_collection], <MilvusException: (code=1, message=Retry run out of 10 retry times, message=collection 442821320644231354 has not been loaded to memory or load failed)>, <Time:{'RPC start': '2023-07-13 07:29:23.413956', 'RPC error': '2023-07-13 07:32:17.572116'}> (decorators.py:108)
[2023-07-13 07:32:17,572 - ERROR - fouram]: RPC error: [load_collection], <MilvusException: (code=1, message=Retry run out of 10 retry times, message=collection 442821320644231354 has not been loaded to memory or load failed)>, <Time:{'RPC start': '2023-07-13 07:29:23.402254', 'RPC error': '2023-07-13 07:32:17.572242'}> (decorators.py:108)
[2023-07-13 07:32:17,573 - ERROR - fouram]: (api_response) : <MilvusException: (code=1, message=Retry run out of 10 retry times, message=collection 442821320644231354 has not been loaded to memory or load failed)> (api_request.py:53)
[2023-07-13 07:32:17,573 - ERROR - fouram]: [CheckFunc] load request check failed, response:<MilvusException: (code=1, message=Retry run out of 10 retry times, message=collection 442821320644231354 has not been loaded to memory or load failed)> (func_check.py:52)
```

memory usage:
![image](https://github.com/milvus-io/milvus/assets/48523564/6572f68f-5336-460f-86eb-8a52c34e1bc4)

### Expected Behavior

load and search successful.

As a comparison test:

Without opening mmap, use the same image: master-20230713-b22d6d2a, insert 16m data and search normally. And the test result is as follows:
```
{'server': {'deploy_tool': 'helm',
            'deploy_mode': 'standalone',
            'config_name': 'standalone_8c16m',
            'config': {'standalone': {'resources': {'limits': {'cpu': '8.0',
                                                               'memory': '16Gi'},
                                                    'requests': {'cpu': '5.0',
                                                                 'memory': '9Gi'}},
                                      'persistence': {'persistentVolumeClaim': {'storageClass': 'local-path'}}},
                       'cluster': {'enabled': False},
                       'etcd': {'replicaCount': 1,
                                'global': {'storageClass': 'local-path'},
                                'metrics': {'enabled': True,
                                            'podMonitor': {'enabled': True}}},
                       'minio': {'mode': 'standalone',
                                 'persistence': {'storageClass': 'local-path'},
                                 'metrics': {'podMonitor': {'enabled': True}}},
                       'pulsar': {'enabled': False},
                       'metrics': {'serviceMonitor': {'enabled': True}},
                       'log': {'level': 'debug'},
                       'image': {'all': {'repository': 'harbor.milvus.io/milvus/milvus',
                                         'tag': 'master-20230713-b22d6d2a'}}},
            'host': 'fouramf-lhl7j-61-6040-milvus.qa-milvus.svc.cluster.local',
            'port': '19530',
            'uri': ''},
 'client': {'test_case_type': 'ConcurrentClientBase',
            'test_case_name': 'test_concurrent_locust_custom_parameters',
            'test_case_params': {'dataset_params': {'metric_type': 'L2',
                                                    'dim': 128,
                                                    'dataset_name': 'sift',
                                                    'dataset_size': '16m',
                                                    'ni_per': 50000},
                                 'collection_params': {},
                                 'load_params': {},
                                 'query_params': {},
                                 'search_params': {},
                                 'resource_groups_params': {'reset': False},
                                 'database_user_params': {'reset_rbac': False,
                                                          'reset_db': False},
                                 'index_params': {'index_type': 'HNSW',
                                                  'index_param': {'M': 8,
                                                                  'efConstruction': 200}},
                                 'concurrent_params': {'concurrent_number': 20,
                                                       'during_time': '10m',
                                                       'interval': 20,
                                                       'spawn_rate': None},
                                 'concurrent_tasks': [{'type': 'search',
                                                       'weight': 20,
                                                       'params': {'nq': 10,
                                                                  'top_k': 10,
                                                                  'search_param': {'ef': 16},
                                                                  'guarantee_timestamp': None,
                                                                  'output_fields': None,
                                                                  'ignore_growing': False,
                                                                  'timeout': 60,
                                                                  'random_data': True}}]},
            'run_id': 2023071392204375,
            'datetime': '2023-07-13 06:20:20.417682',
            'client_version': '2.2'},
 'result': {'test_result': {'index': {'RT': 1347.962},
                            'insert': {'total_time': 1432.3291,
                                       'VPS': 11170.6171,
                                       'batch_time': 4.476,
                                       'batch': 50000},
                            'flush': {'RT': 14.5353},
                            'load': {'RT': 24.387},
                            'Locust': {'Aggregated': {'Requests': 149008,
                                                      'Fails': 0,
                                                      'RPS': 248.31,
                                                      'fail_s': 0.0,
                                                      'RT_max': 512.55,
                                                      'RT_avg': 64.93,
                                                      'TP50': 65,
                                                      'TP99': 120.0},
                                       'search': {'Requests': 149008,
                                                  'Fails': 0,
                                                  'RPS': 248.31,
                                                  'fail_s': 0.0,
                                                  'RT_max': 512.55,
                                                  'RT_avg': 64.93,
                                                  'TP50': 65,
                                                  'TP99': 120.0}}}}}
```

### Steps To Reproduce

```markdown
1. create a collection or use an existing collection
        2. build index on vector column => HNSW
        3. insert a certain number of vectors  => 16m
        4. flush collection
        5. build index on vector column with the same parameters
        6. build index on on scalars column or not
        7. count the total number of rows
        8. load collection
        9. perform concurrent operations
        10. clean all collections or not
```

### Anything else?

```
{'standalone': {'resources': {'limits': {'cpu': '8.0', 'memory': '16Gi'}, 'requests': {'cpu': '5.0', 'memory': '9Gi'}}, 'persistence': {'persistentVolumeClaim': {'storageClass': 'local-path'}}}, 'cluster': {'enabled': False}, 'etcd': {'replicaCount': 1, 'global': {'storageClass': 'local-path'}, 'metrics': {'enabled': True, 'podMonitor': {'enabled': True}}}, 'minio': {'mode': 'standalone', 'persistence': {'storageClass': 'local-path'}, 'metrics': {'podMonitor': {'enabled': True}}}, 'pulsar': {'enabled': False}, 'metrics': {'serviceMonitor': {'enabled': True}}, 'log': {'level': 'debug'}, 'extraConfigFiles': {'user.yaml': 'queryNode:\n  mmapDirPath: /mmap\n'}, 'image': {'all': {'repository': 'harbor.milvus.io/milvus/milvus', 'tag': 'master-20230713-b22d6d2a'}}}, 'standalone_8c16m', {'standalone': {'resources': {'limits': {'cpu': '8.0', 'memory': '16Gi'}, 'requests': {'cpu': '5.0', 'memory': '9Gi'}}, 'persistence': {'persistentVolumeClaim': {'storageClass': 'local-path'}}}, 'cluster': {'enabled': False}, 'etcd': {'replicaCount': 1, 'global': {'storageClass': 'local-path'}, 'metrics': {'enabled': True, 'podMonitor': {'enabled': True}}}, 'minio': {'mode': 'standalone', 'persistence': {'storageClass': 'local-path'}, 'metrics': {'podMonitor': {'enabled': True}}}, 'pulsar': {'enabled': False}, 'metrics': {'serviceMonitor': {'enabled': True}}, 'log': {'level': 'debug'}, 'extraConfigFiles': {'user.yaml': 'queryNode:\n  mmapDirPath: /mmap\n'}, 'image': {'all': {'repository': 'harbor.milvus.io/milvus/milvus', 'tag': 'master-20230713-b22d6d2a'}}} 
```

**Comment by yah01** [comment]
the pod got OOM killed, working on it to determine whether it's normal

**Comment by yah01** [comment]
The system got OOM while building index, the load stage had not started yet

**Comment by yah01** [comment]
The memory overhead of loading index is still too high

**Comment by xiaofan-luan** [comment]
> The memory overhead of loading index is still too high

So we should think of do the index assemble on disk if any?

**Comment by yah01** [comment]
> > The memory overhead of loading index is still too high
> 
> 
> 
> So we should think of do the index assemble on disk if any?

I'm going to change the loading index process to be like loading segment, streaming loading with a buffer then we can limit the memory usage by adjust the buffer size. For mmap, the result will be written directly to disk

**Comment by yah01** [comment]
/assign @elstic 
fixed with #25698

**Comment by elstic** [comment]
The issue has been fixed, 
verify the image: master-20230724-8e5d6193
